crypto/curve25519/curve25519.c

```diff
@@ -341,23 +335,51 @@ int ED25519_verify(const uint8_t *message, size_t message_len,
     }
   }
 
+  // Decode public key as A'.
+  ge_p3 A;
+  if (!x25519_ge_frombytes_vartime(&A, public_key)) {
+    return 0;
+  }
+  uint8_t pkcopy[32];
+  OPENSSL_memcpy(pkcopy, public_key, 32);
+
+  // Step: rfc8032 5.1.7.2
+  // Compute SHA512(dom2(F,C) || R || public_key || PH(message)).
+  // For Ed25519, dom2(F,C) is the empty string, cf. rfc8032 5.1.
   SHA512_CTX hash_ctx;
   SHA512_Init(&hash_ctx);
   SHA512_Update(&hash_ctx, signature, 32);
   SHA512_Update(&hash_ctx, public_key, 32);
   SHA512_Update(&hash_ctx, message, message_len);
-  uint8_t h[SHA512_DIGEST_LENGTH];
-  SHA512_Final(h, &hash_ctx);
+  uint8_t k[SHA512_DIGEST_LENGTH];
+  SHA512_Final(k, &hash_ctx);
+
+  // Reduce k modulo the order of the base-point B. Saves compute in the
+  // subsequent scalar multiplication.
+  x25519_sc_reduce(k);
 
-  x25519_sc_reduce(h);
+  // Step: rfc8032 5.1.7.3
+  // We will check [S]B - [k]A' =? R_expected.
+
+  // First negate A'. Point negation for the twisted edwards curve when points
+  // are represented in the extended coordinate system is simply:
+  //   -(X,Y,Z,T) = (-X,Y,Z,-T).
+  // See "Twisted Edwards curves revisited" https://ia.cr/2008/522.
+  fe_loose t;
+  fe_neg(&t, &A.X);
+  fe_carry(&A.X, &t);
+  fe_neg(&t, &A.T);
+  fe_carry(&A.T, &t);
 
-  ge_p2 R;
-  ge_double_scalarmult_vartime(&R, h, &A, scopy);
+  // Compute R_have <- [S]B - [k]A'.
+  ge_p2 R_have;
+  ge_double_scalarmult_vartime(&R_have, k, &A, S);
 
-  uint8_t rcheck[32];
-  x25519_ge_tobytes(rcheck, &R);
+  uint8_t R_have_enc[32];
```

Comment:
Maybe add a comment here or rename it to `R_have_encoded` since "enc" can be read as "encrypted".

Reply:
makes sense!